src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c

```diff
@@ -26,17 +51,157 @@ const ActorInit Shot_Sun_InitVars = {
+    osSyncPrintf("%d ---- オカリナの秘密発生!!!!!!!!!!!!!\n", this->actor.params);
+    params = this->actor.params & 0xFF;
+    if (params == 0x40 || params == 0x41) {
+        this->unk_1A4[0] = 0;
+        this->actor.flags |= 0x10;
+        this->actor.flags |= 0x2000000;
+        this->actionFunc = func_80BADF0C;
+        this->actor.flags |= 0x8000000;
+    } else {
+        Collider_InitCylinder(globalCtx, &this->collider);
+        Collider_SetCylinder(globalCtx, &this->collider, &this->actor, &sCylinderInit);
+        this->actionFunc = ShotSun_HyliaSunUpdate;
+        this->actor.flags &= ~1;
+    }
+}
+
+void ShotSun_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    ShotSun* this = THIS;
+    s32 params = this->actor.params & 0xFF;
+
+    if (params != 0x40 && params != 0x41) {
+        Collider_DestroyCylinder(globalCtx, &this->collider);
+    }
+}
+
+void ShotSun_FairyCountdown(ShotSun* this, GlobalContext* globalCtx) {
+    s32 params = this->actor.params & 0xFF;
+    s32 fairyParams;
+
+    if (this->timer > 0) {
+        this->timer--;
+        return;
+    }
+
+    switch (params) {
+        case 0x40:
+            fairyParams = 7;
+            break;
+        case 0x41:
+            fairyParams = 7;
+            break;
+    }
+
+    Actor_Spawn(&globalCtx->actorCtx, globalCtx, ACTOR_EN_ELF, this->actor.initPosRot.pos.x,
+                this->actor.initPosRot.pos.y, this->actor.initPosRot.pos.z, 0, 0, 0, fairyParams);
+
+    Actor_Kill(&this->actor);
+}
+
+void ShotSun_StartFairyCountdown(ShotSun* this, GlobalContext* globalCtx) {
+    if ((func_8005B198() == this->actor.type) || (this->timer != 0)) {
+        this->actionFunc = ShotSun_FairyCountdown;
+        this->timer = 50;
+
+        Actor_Spawn(&globalCtx->actorCtx, globalCtx, ACTOR_DEMO_KANKYO, this->actor.initPosRot.pos.x,
+                    this->actor.initPosRot.pos.y, this->actor.initPosRot.pos.z, 0, 0, 0, 0x11);
+
+        func_80078914(&this->actor.projectedPos, NA_SE_EV_TRE_BOX_APPEAR);
+    }
+}
+
+void func_80BADF0C(ShotSun* this, GlobalContext* globalCtx) {
+    Player* player = PLAYER;
+    s32 pad;
+    s32 params = this->actor.params & 0xFF;
+
+    if (func_800CB650(&this->actor.posRot.pos, &player->actor.posRot.pos) > 22500.0f) {
+        this->unk_1A4[0] = 0;
+    } else {
+        if (this->unk_1A4[0] == 0) {
+            if (!(player->stateFlags2 & 0x1000000)) {
+                player->stateFlags2 |= 0x800000;
+                return;
+            } else {
+                this->unk_1A4[0] = 1;
+            }
+        }
+        if (this->unk_1A4[0] == 1) {
+            func_8010BD58(globalCtx, 1U);
+            this->unk_1A4[0] = 2;
+            return;
+        }
+
+        if (this->unk_1A4[0] == 2 && globalCtx->msgCtx.unk_E3EE == 4) {
+            if (((params == 0x40) && (*(u16*)(&globalCtx->msgCtx.unk_E3E8[4]) == 9)) ||
+                ((params == 0x41) && (*(u16*)(&globalCtx->msgCtx.unk_E3E8[4]) == 0xB))) {
```

Comment:
is all of this pointer stuff and dereferencing needed? you should be able to just access those values normally

Reply:
Yes, here it's loading a 16 bit value from unknown data left as a char array in the message context so I need to do this unless a proper field is put on the message context.

Reply:
if you know the type then you can set the msgContext variables up correctly. would be much preferred to do that instead

Reply:
Got it.